src/models/transaction.rs

```diff
@@ -1,197 +1,161 @@
+    let mut signed_data = Vec::with_capacity(transaction_signed.transaction.length());
+    transaction_signed.transaction.encode_without_signature(&mut signed_data);
+
+    // Pack the calldata in 31-byte chunks
+    let mut signed_data: Vec<FieldElement> = std::iter::once(FieldElement::from(signed_data.len()))
+        .chain(signed_data.chunks(31).filter_map(|chunk_bytes| FieldElement::from_byte_slice_be(chunk_bytes).ok()))
+        .collect();
+
+    // Prepare the calldata for the Starknet invoke transaction
+    let capacity = 6 + signed_data.len();
+
+    // Check if call data is too large
+    #[cfg(not(feature = "hive"))]
+    if capacity > *MAX_FELTS_IN_CALLDATA {
+        return Err(EthApiError::CalldataExceededLimit(*MAX_FELTS_IN_CALLDATA, capacity));
     }
 
-    test_rpc_to_primitive_conversion!(
-        test_legacy_transaction_conversion_to_primitive,
-        legacy_rpc_transaction,
-        gas_price,
-        false
-    );
-
-    test_rpc_to_ec_recovered_conversion!(
-        test_legacy_transaction_conversion_to_ec_recovered,
-        legacy_rpc_transaction,
-        gas_price,
-        false
-    );
-
-    test_rpc_to_primitive_conversion!(
-        test_eip2930_transaction_conversion_to_primitive,
-        eip2930_rpc_transaction,
-        gas_price,
-        true
-    );
-    test_rpc_to_ec_recovered_conversion!(
-        test_eip2930_transaction_conversion_to_ec_recovered,
-        eip2930_rpc_transaction,
-        gas_price,
-        true
-    );
-
-    test_rpc_to_primitive_conversion!(
-        test_eip1559_transaction_conversion_to_primitive,
-        eip1559_rpc_transaction,
-        max_fee_per_gas,
-        true
-    );
-
-    test_rpc_to_ec_recovered_conversion!(
-        test_eip1559_transaction_conversion_to_ec_recovered,
-        eip1559_rpc_transaction,
-        max_fee_per_gas,
-        true
-    );
+    let mut calldata = Vec::with_capacity(capacity);
+
+    // assert that the selector < FieldElement::MAX - retries
+    assert!(*ETH_SEND_TRANSACTION < FieldElement::MAX - retries.into());
+    let selector = *ETH_SEND_TRANSACTION + retries.into();
+
+    // Retries are used to alter the transaction hash in order to avoid the
+    // DuplicateTx error from the Starknet gateway, encountered whenever
+    // a transaction with the same hash is sent multiple times.
+    // We add the retries to the selector in the calldata, since the selector
+    // is not used in by the EOA contract during the transaction execution.
+    calldata.append(&mut vec![
+        FieldElement::ONE,        // call array length
+        *KAKAROT_ADDRESS,         // contract address
+        selector,                 // selector + retries
+        FieldElement::ZERO,       // data offset
+        signed_data.len().into(), // data length
+        signed_data.len().into(), // calldata length
+    ]);
+    calldata.append(&mut signed_data);
+
+    Ok(calldata)
+}
+
+pub(crate) async fn transaction_retries(
```

Comment:
What do you think about migrating this function to the `Database` structure to minimize functions which are not inside objects? 

So that we can call: `database.transaction_retries(transaction_signed)`